cli/src/commands/diff.rs

```diff
@@ -0,0 +1,496 @@
+    pallet_metadata_2: &'a PalletMetadata<'a>,
+) -> Vec<Diff<&'a StorageEntryMetadata>> {
+    let mut storage_entries: HashMap<
+        &str,
+        (
+            Option<&'a StorageEntryMetadata>,
+            Option<&'a StorageEntryMetadata>,
+        ),
+    > = HashMap::new();
+    if let Some(storage_metadata) = pallet_metadata_1.storage() {
+        for storage_entry in storage_metadata.entries() {
+            let (e1, _) = storage_entries.entry(storage_entry.name()).or_default();
+            *e1 = Some(storage_entry);
+        }
+    }
+    if let Some(storage_metadata) = pallet_metadata_2.storage() {
+        for storage_entry in storage_metadata.entries() {
+            let (e1, e2) = storage_entries.entry(storage_entry.name()).or_default();
+            // skip all entries with the same hash:
+            if let Some(e1_inner) = e1 {
+                let e1_hash = pallet_metadata_1
+                    .storage_hash(e1_inner.name())
+                    .expect("storage entry should be present");
+                let e2_hash = pallet_metadata_2
+                    .storage_hash(storage_entry.name())
+                    .expect("storage entry should be present");
+                if e1_hash == e2_hash {
+                    storage_entries.remove(storage_entry.name());
+                    continue;
+                }
+            }
+            *e2 = Some(storage_entry);
+        }
+    }
+    storage_entries
+        .into_values()
+        .map(|tuple| Diff::try_from(tuple).unwrap())
+        .collect()
+}
+
+type CallsHashmap<'a> = HashMap<
+    &'a str,
+    (
+        Option<&'a Variant<PortableForm>>,
+        Option<&'a Variant<PortableForm>>,
+    ),
+>;
+
+fn calls_differences<'a>(
+    pallet_metadata_1: &'a PalletMetadata<'a>,
+    pallet_metadata_2: &'a PalletMetadata<'a>,
+) -> Vec<Diff<&'a Variant<PortableForm>>> {
+    let mut calls: CallsHashmap = HashMap::new();
+    if let Some(call_variants) = pallet_metadata_1.call_variants() {
+        for call_variant in call_variants {
+            let (e1, _) = calls.entry(&call_variant.name).or_default();
+            *e1 = Some(call_variant);
+        }
+    }
+    if let Some(call_variants) = pallet_metadata_2.call_variants() {
+        for call_variant in call_variants {
+            let (e1, e2) = calls.entry(&call_variant.name).or_default();
+            // skip all entries with the same hash:
+            if let Some(e1_inner) = e1 {
+                let e1_hash = pallet_metadata_1
+                    .call_hash(&e1_inner.name)
+                    .expect("call should be present");
+                let e2_hash = pallet_metadata_2
+                    .call_hash(&call_variant.name)
+                    .expect("call should be present");
+                if e1_hash == e2_hash {
+                    calls.remove(call_variant.name.as_str());
+                    continue;
+                }
+            }
+            *e2 = Some(call_variant);
+        }
+    }
```

Comment:
The general pattern of "give me a list of the differences between two things" feels like something that could be abstracted here since it's repeated several times; the only logic that changes each time is "how do I turn a thing into something comparable". 

I'd propose it'd be a fun challenge to have a go at doing this abstracting; I could imagine ultimately ending with a signature something like this:

```rust
fn diff<T, C: PartialEq>(
    items_a: impl IntoIterator<Item = T>,
    comparable_a: impl Fn(&T) -> C
    items_b: impl IntoIterator<Item = T>,
    comparable_b: impl Fn(&T) -> C
) -> Vec<Diff<T>>
```